# Phase 2

### Data transformation

### 1. Read the data from the staging area and find outliers, null or empty data, and more.

In [1]:
import os
from dotenv import load_dotenv
import psycopg2
from sqlalchemy import create_engine, text
import pandas as pd

# Load environment variables
load_dotenv()

db_staging_user = os.getenv('DB_STAGING_USER')
db_staging_password = os.getenv('DB_STAGING_PASSWORD')
db_staging_host = os.getenv('DB_STAGING_HOST')
db_staging_port = int(os.getenv('DB_STAGING_PORT', 5432))
db_staging_name = os.getenv('DB_STAGING_NAME')

resources_path = os.getenv('RESOURCES_PATH')
mascotas_propietarios_filename = 'Mascotas_Propietarios_despensaAnimal_Generated.csv'
propietarios_transacciones_filename = 'Propietarios_Transacciones_despensaAnimal_Generated.csv'

if db_staging_user is None:
    raise ValueError('DB_STAGING_USER is not set')
if db_staging_password is None:
    raise ValueError('DB_STAGING_PASSWORD is not set')
if db_staging_host is None:
    raise ValueError('DB_STAGING_HOST is not set')
if db_staging_port is None:
    raise ValueError('DB_STAGING_PORT is not set')
if db_staging_name is None:
    raise ValueError('DB_STAGING_NAME is not set')
if resources_path is None:
    raise ValueError('RESOURCES_PATH is not set')

connStaging = psycopg2.connect(
    dbname=db_staging_name,
    user=db_staging_user,
    password=db_staging_password,
    host=db_staging_host,
    port=db_staging_port
)
connStaging.autocommit = True   

#### 1.1 Create the database engine for the staging data

In [2]:
db_staging_engine = create_engine(f'postgresql://{db_staging_user}:{db_staging_password}@{db_staging_host}:{db_staging_port}/{db_staging_name}')

df_mascotas_propietarios_staging = pd.read_sql('SELECT * FROM mascotas_propietarios_staging;', db_staging_engine)
df_mascotas_propietarios_staging

nombre_mascota                raza   peso fecha_nacimiento    sexo  \
0       THANOS URREGO              Poodle   None       2022-09-01   Macho   
1      TINENK ORDONEZ             Shit-zu   None       2017-02-17  Hembra   
2         SIMON VACCA                 PUG   None       2021-08-01   Macho   
3        SUSSY PUERTA  YORK SHIRE TERRIER    3.6       2016-10-07  Hembra   
4       EEVEE ANACONA              Poodle   None       2017-09-03  Hembra   
...               ...                 ...    ...              ...     ...   
11994        Jennifer    Golden Retriever   46.1       2024-07-17  Hembra   
11995         Timothy     German Shepherd  41.79       2014-03-10   Macho   
11996          Marcus    Golden Retriever  32.76       2014-09-16  Hembra   
11997           James              Poodle  35.41       2005-08-23  Hembra   
11998         Timothy            Labrador   25.4       2005-09-11   Macho   

      temperamento numero_carnet estado_reproductivo numero_partos   color  \
0           social          1631              Entero          None    None   
1           social          1650                None          None    None   
2           social          1184                None          None    None   
3           social          1359              ENTERA       NINGUNO    None   
4           social          1632                None          None    None   
...            ...           ...                 ...           ...     ...   
11994     Agresivo      FB13AD2F        Esterilizado             1   Negro   
11995     Agresivo      2361BF30     No Esterilizado             0  Blanco   
11996     Sociable      85D44DE5        Esterilizado             0  Marrón   
11997     Sociable      4C95A084     No Esterilizado             5    Gris   
11998     Sociable      CD183035     No Esterilizado             0   Negro   

       ...            ciudad  \
0      ...              Cali   
1      ...  Santiago de Cali   
2      ...              None   
3      ...              None   
4      ...              Cali   
...    ...               ...   
11994  ...              Cali   
11995  ...              Cali   
11996  ...              Cali   
11997  ...              Cali   
11998  ...              Cali   

                                               direccion    telefono  \
0      CRA100 #28-68 VALLE DEL LILI  MADEIRO TORRE 1 ...        None   
1                              cra 85 c # 33- 40 casa 54  3058147632   
2                              Cra98B #45-200 SAN MIGUEL        None   
3                   Cra. 98b #34-53 GUADALQUIVIR CASA 64  3134024437   
4                      Calle 45#98B-65 Apto 403, torre 8        None   
...                                                  ...         ...   
11994                                   2256 Daniel Camp  3139761188   
11995                       194 Kimberly Tunnel Apt. 031  3184965612   
11996                                  237 Castro Meadow  3168512890   
11997                                  6386 Miller Place  3201628333   
11998                                   20399 Tran Ports  3001442040   

         whatsapp                       email tipo_documento numero_documento  \
0      3207201137  lizethurrego1990@gmail.com             CC       1036637677   
1      3157929392      pa_ordonez@hotmail.com             CC         66987417   
2      3234191060                        None             CC             None   
3      3146096191                        None             CC       1006107262   
4      3216066041        jazmin.dag@gmail.com             CC       1061750508   
...           ...                         ...            ...              ...   
11994        None         jacob96@example.com            NIT        176580896   
11995        None   shelleyparker@example.com            NIT        130286470   
11996        None        ojohnson@example.org             CE        194581954   
11997        None     jasminecarr@example.com            NIT        174849231   
1199

### 2. Remove data that not add value to the transacctions scenario

1.	comentarios_fallecimiento → Not useful for purchases or analysis of services.
2.	motivo_fallecimiento → Not related to transactions.
3.	fecha_fallecimiento → If the pet has died, it will no longer generate new purchases.
4.	numero_carnet → Not relevant to link purchases with clients.
5.	estado_reproductivo → Does not influence the purchase of services.
6.	numero_partos → Does not seem to affect the purchase of veterinary services.
7.	profesion → Not directly related to service purchases.
8.	notificaciones_whatsapp → Not used to analyze transactions.

In [3]:
# Function to drop unnecessary columns
def drop_unnecessary_columns(df: pd.DataFrame) -> pd.DataFrame:
    columns_to_drop = [
        'comentarios_fallecimiento',
        'motivo_fallecimiento',
        'fecha_fallecimiento',
        'numero_carnet',
        'estado_reproductivo',
        'numero_partos',
        'profesion',
        'notificaciones_whatsapp'
    ]

    for col in columns_to_drop:
        if col in df.columns:
            print(f'Removing column: {col}')
            df.drop(columns=[col], inplace=True)
        else:
            print(f'Warning: Column {col} not found in the dataset.')
    
    print(f'Columns removed. New shape: {df.shape}')
    return df

# Apply the function
df_mascotas_propietarios_cleaned = drop_unnecessary_columns(df_mascotas_propietarios_staging)
df_mascotas_propietarios_cleaned

Removing column: comentarios_fallecimiento
Removing column: motivo_fallecimiento
Removing column: fecha_fallecimiento
Removing column: numero_carnet
Removing column: estado_reproductivo
Removing column: numero_partos
Removing column: profesion
Removing column: notificaciones_whatsapp
Columns removed. New shape: (11999, 16)


nombre_mascota                raza   peso fecha_nacimiento    sexo  \
0       THANOS URREGO              Poodle   None       2022-09-01   Macho   
1      TINENK ORDONEZ             Shit-zu   None       2017-02-17  Hembra   
2         SIMON VACCA                 PUG   None       2021-08-01   Macho   
3        SUSSY PUERTA  YORK SHIRE TERRIER    3.6       2016-10-07  Hembra   
4       EEVEE ANACONA              Poodle   None       2017-09-03  Hembra   
...               ...                 ...    ...              ...     ...   
11994        Jennifer    Golden Retriever   46.1       2024-07-17  Hembra   
11995         Timothy     German Shepherd  41.79       2014-03-10   Macho   
11996          Marcus    Golden Retriever  32.76       2014-09-16  Hembra   
11997           James              Poodle  35.41       2005-08-23  Hembra   
11998         Timothy            Labrador   25.4       2005-09-11   Macho   

      temperamento   color   nombre_propietario            ciudad  \
0           social    None        LIZETH URREGO              Cali   
1           social    None        PAOLA ORDONEZ  Santiago de Cali   
2           social    None         ANDREA VACCA              None   
3           social    None  MARIA CAMILA PUERTA              None   
4           social    None       YAZMIN ANACONA              Cali   
...            ...     ...                  ...               ...   
11994     Agresivo   Negro        Steven Keller              Cali   
11995     Agresivo  Blanco    Christopher Clark              Cali   
11996     Sociable  Marrón           Lisa Reyes              Cali   
11997     Sociable    Gris           Paul White              Cali   
11998     Sociable   Negro        Deborah Lopez              Cali   

                                               direccion    telefono  \
0      CRA100 #28-68 VALLE DEL LILI  MADEIRO TORRE 1 ...        None   
1                              cra 85 c # 33- 40 casa 54  3058147632   
2                              Cra98B #45-200 SAN MIGUEL        None   
3                   Cra. 98b #34-53 GUADALQUIVIR CASA 64  3134024437   
4                      Calle 45#98B-65 Apto 403, torre 8        None   
...                                                  ...         ...   
11994                                   2256 Daniel Camp  3139761188   
11995                       194 Kimberly Tunnel Apt. 031  3184965612   
11996                                  237 Castro Meadow  3168512890   
11997                                  6386 Miller Place  3201628333   
11998                                   20399 Tran Ports  3001442040   

         whatsapp                       email tipo_documento numero_documento  \
0      3207201137  lizethurrego1990@gmail.com             CC       1036637677   
1      3157929392      pa_ordonez@hotmail.com             CC         66987417   
2      3234191060                        None             CC             None   
3      3146096191                        None             CC       1006107262   
4      3216066041        jazmin.dag@gmail.com             CC       1061750508   
...           ...                         ...            ...              ...   
11994        None         jacob96@example.com            NIT        176580896   
11995        None   shelleyparker@example.com            NIT        130286470   
11996        None        ojohnson@example.org             CE        194581954   
11997        None     jasminecarr@example.com            NIT        174849231   
11998        None         sarah84@example.org             CC        146255830   

         estado  
0        Activo  
1        Activo  
2        Activo  
3        Activo  
4        Activo  
...         ...  
11994  Inactivo  
11995    Activo  
11996    Activo  
11997  Inactivo  
11998  Inactivo  

[11999 rows x 16 columns]

### 2.1 Analyze the field: ciudad and its different values

In [4]:
df_ciudad = df_mascotas_propietarios_cleaned.groupby("ciudad").size().reset_index(name="total_registros")
df_ciudad

ciudad  total_registros
0                                             CAI                1
1                                            CALI             1692
2                                           CALUI                1
3                                  CRA 64 # 14-24                1
4                                            Cali             9259
5   Calle 60 b # 119 - 47 Torre 2 apto 602 Laurel                1
6                                            EEUU                5
7                                         JAMUNDI                8
8                                Santiago de Cali              555
9                                            cali               54
10            calle 50 # 99A-66 bloque 5 apto 203                1

We can see that the field: 'ciudad' doesn't have the same value and, also has information relates with countries, others towns and addresses, we must clean this data.

In [5]:
# Update all values in 'ciudad' column to 'Cali'
df_mascotas_propietarios_cleaned["ciudad"] = "Cali"

df_ciudad = df_mascotas_propietarios_cleaned.groupby("ciudad").size().reset_index(name="total_registros")
df_ciudad

ciudad  total_registros
0   Cali            11999

### 2.2 Analyze the field: peso and its different values

In [6]:
df_mascotas_propietarios_cleaned

nombre_mascota                raza   peso fecha_nacimiento    sexo  \
0       THANOS URREGO              Poodle   None       2022-09-01   Macho   
1      TINENK ORDONEZ             Shit-zu   None       2017-02-17  Hembra   
2         SIMON VACCA                 PUG   None       2021-08-01   Macho   
3        SUSSY PUERTA  YORK SHIRE TERRIER    3.6       2016-10-07  Hembra   
4       EEVEE ANACONA              Poodle   None       2017-09-03  Hembra   
...               ...                 ...    ...              ...     ...   
11994        Jennifer    Golden Retriever   46.1       2024-07-17  Hembra   
11995         Timothy     German Shepherd  41.79       2014-03-10   Macho   
11996          Marcus    Golden Retriever  32.76       2014-09-16  Hembra   
11997           James              Poodle  35.41       2005-08-23  Hembra   
11998         Timothy            Labrador   25.4       2005-09-11   Macho   

      temperamento   color   nombre_propietario ciudad  \
0           social    None        LIZETH URREGO   Cali   
1           social    None        PAOLA ORDONEZ   Cali   
2           social    None         ANDREA VACCA   Cali   
3           social    None  MARIA CAMILA PUERTA   Cali   
4           social    None       YAZMIN ANACONA   Cali   
...            ...     ...                  ...    ...   
11994     Agresivo   Negro        Steven Keller   Cali   
11995     Agresivo  Blanco    Christopher Clark   Cali   
11996     Sociable  Marrón           Lisa Reyes   Cali   
11997     Sociable    Gris           Paul White   Cali   
11998     Sociable   Negro        Deborah Lopez   Cali   

                                               direccion    telefono  \
0      CRA100 #28-68 VALLE DEL LILI  MADEIRO TORRE 1 ...        None   
1                              cra 85 c # 33- 40 casa 54  3058147632   
2                              Cra98B #45-200 SAN MIGUEL        None   
3                   Cra. 98b #34-53 GUADALQUIVIR CASA 64  3134024437   
4                      Calle 45#98B-65 Apto 403, torre 8        None   
...                                                  ...         ...   
11994                                   2256 Daniel Camp  3139761188   
11995                       194 Kimberly Tunnel Apt. 031  3184965612   
11996                                  237 Castro Meadow  3168512890   
11997                                  6386 Miller Place  3201628333   
11998                                   20399 Tran Ports  3001442040   

         whatsapp                       email tipo_documento numero_documento  \
0      3207201137  lizethurrego1990@gmail.com             CC       1036637677   
1      3157929392      pa_ordonez@hotmail.com             CC         66987417   
2      3234191060                        None             CC             None   
3      3146096191                        None             CC       1006107262   
4      3216066041        jazmin.dag@gmail.com             CC       1061750508   
...           ...                         ...            ...              ...   
11994        None         jacob96@example.com            NIT        176580896   
11995        None   shelleyparker@example.com            NIT        130286470   
11996        None        ojohnson@example.org             CE        194581954   
11997        None     jasminecarr@example.com            NIT        174849231   
11998        None         sarah84@example.org             CC        146255830   

         estado  
0        Activo  
1        Activo  
2        Activo  
3        Activo  
4        Activo  
...         ...  
11994  Inactivo  
11995    Activo  
11996    Activo  
11997  Inactivo  
11998  Inactivo  

[11999 rows x 16 columns]

We can see that the field: 'peso' mix numeric and string values like 9.2Kg. We must change that to only numeric values.

In [7]:
df_mascotas_propietarios_cleaned.groupby("peso").size().reset_index(name="total_registros")

peso  total_registros
0           0.4                1
1          0.53                1
2           0.6                3
3        0.6 Kg                1
4           0.7                1
...         ...              ...
4899        9KG                1
4900        9Kg                1
4901        9kg                1
4902       9kg                 1
4903  PENDIENTE                1

[4904 rows x 2 columns]

In [8]:
# Convert all values in 'peso' column to float, if possible (otherwise, set to NaN)
df_mascotas_propietarios_cleaned["peso"] = pd.to_numeric(df_mascotas_propietarios_cleaned["peso"], errors='coerce')
df_mascotas_propietarios_cleaned.groupby("peso").size().reset_index(name="total_registros")


peso  total_registros
0       0.40                1
1       0.53                1
2       0.60                3
3       0.70                1
4       1.00                3
...      ...              ...
4668   59.98                2
4669   59.99                1
4670   60.00                2
4671   64.00                1
4672  555.00                1

[4673 rows x 2 columns]

### 2.3 Analyze the field: 'date' and its different values

In [9]:
df_mascotas_propietarios_cleaned.groupby("fecha_nacimiento").size().reset_index(name="total_registros")

fecha_nacimiento  total_registros
0          0204-02-21                1
1          2002-05-18                1
2          2002-11-09                1
3          2003-02-23                1
4          2003-09-26                1
...               ...              ...
5728       2025-03-03                1
5729       2025-03-04                4
5730       2025-03-05                2
5731       2025-03-07                3
5732       2025-03-08                2

[5733 rows x 2 columns]

We can see that the field: 'fecha_nacimiento' has wrong date values like 0204-02-21. That date is wrong because the business
was created in the year 2016.

In [10]:
# Convert all values in 'fecha_nacimiento' column to a datetime greater than 2016-01-01
df_mascotas_propietarios_cleaned["fecha_nacimiento"] = pd.to_datetime(df_mascotas_propietarios_cleaned["fecha_nacimiento"], errors='coerce')
df_mascotas_propietarios_cleaned = df_mascotas_propietarios_cleaned[df_mascotas_propietarios_cleaned["fecha_nacimiento"] >= '2016-01-01']
df_mascotas_propietarios_cleaned.groupby("fecha_nacimiento").size().reset_index(name="total_registros")

fecha_nacimiento  total_registros
0          2016-01-01                7
1          2016-01-02                3
2          2016-01-03                1
3          2016-01-04                2
4          2016-01-05                1
...               ...              ...
2646       2025-03-03                1
2647       2025-03-04                4
2648       2025-03-05                2
2649       2025-03-07                3
2650       2025-03-08                2

[2651 rows x 2 columns]

### 2.4 Analyze the field: 'telefono' and 'whatsapp'

In [11]:
# Ensure 'telefono' column only contains numeric values (otherwise, set to None)
df_mascotas_propietarios_cleaned.loc[:, "telefono"] = df_mascotas_propietarios_cleaned["telefono"].apply(
    lambda x: x if str(x).isnumeric() else None
)

# Fill missing 'telefono' values with 'whatsapp' numbers
df_mascotas_propietarios_cleaned.loc[:, "telefono"] = df_mascotas_propietarios_cleaned["telefono"].fillna(
    df_mascotas_propietarios_cleaned["whatsapp"]
)

# Display the updated DataFrame
df_mascotas_propietarios_cleaned[["telefono", "whatsapp"]]

# Drop the 'whatsapp' column
df_mascotas_propietarios_cleaned.drop(columns=["whatsapp"], inplace=True)
df_mascotas_propietarios_cleaned

/var/folders/v7/3zqcfbbd54zbnvx9px7f533h0000gn/T/ipykernel_46647/689292948.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_mascotas_propietarios_cleaned.drop(columns=["whatsapp"], inplace=True)


nombre_mascota                raza   peso fecha_nacimiento    sexo  \
0       THANOS URREGO              Poodle    NaN       2022-09-01   Macho   
1      TINENK ORDONEZ             Shit-zu    NaN       2017-02-17  Hembra   
2         SIMON VACCA                 PUG    NaN       2021-08-01   Macho   
3        SUSSY PUERTA  YORK SHIRE TERRIER   3.60       2016-10-07  Hembra   
4       EEVEE ANACONA              Poodle    NaN       2017-09-03  Hembra   
...               ...                 ...    ...              ...     ...   
11987         William              Poodle  11.05       2017-04-12  Hembra   
11988            Anna    Golden Retriever   5.08       2024-10-28   Macho   
11990          Andrew     German Shepherd  25.12       2016-08-23   Macho   
11991         Charles              Poodle  47.89       2017-12-10  Hembra   
11994        Jennifer    Golden Retriever  46.10       2024-07-17  Hembra   

      temperamento  color   nombre_propietario ciudad  \
0           social   None        LIZETH URREGO   Cali   
1           social   None        PAOLA ORDONEZ   Cali   
2           social   None         ANDREA VACCA   Cali   
3           social   None  MARIA CAMILA PUERTA   Cali   
4           social   None       YAZMIN ANACONA   Cali   
...            ...    ...                  ...    ...   
11987     Juguetón  Negro         Angela Moore   Cali   
11988    Tranquilo  Negro       Steven Douglas   Cali   
11990    Tranquilo   Gris         Debbie Smith   Cali   
11991       Tímido   Gris            Jose Reid   Cali   
11994     Agresivo  Negro        Steven Keller   Cali   

                                               direccion    telefono  \
0      CRA100 #28-68 VALLE DEL LILI  MADEIRO TORRE 1 ...  3207201137   
1                              cra 85 c # 33- 40 casa 54  3058147632   
2                              Cra98B #45-200 SAN MIGUEL  3234191060   
3                   Cra. 98b #34-53 GUADALQUIVIR CASA 64  3134024437   
4                      Calle 45#98B-65 Apto 403, torre 8  3216066041   
...                                                  ...         ...   
11987                                  0106 Dudley Route  3143111203   
11988                       6458 Rogers Courts Suite 080  3191768598   
11990                            80840 Sellers Junctions  3118659393   
11991                              7585 Rodriguez Bypass  3024903661   
11994                                   2256 Daniel Camp  3139761188   

                            email tipo_documento numero_documento    estado  
0      lizethurrego1990@gmail.com             CC       1036637677    Activo  
1          pa_ordonez@hotmail.com             CC         66987417    Activo  
2                            None             CC             None    Activo  
3                            None             CC       1006107262    Activo  
4            jazmin.dag@gmail.com             CC       1061750508    Activo  
...                           ...            ...              ...       ...  
11987  hardingkenneth@example.com            NIT        125606839  Inactivo  
11988     jeremyhuber@example.net             CE        176912006    Activo  
11990       brandon31@example.net             CC        195453083  Inactivo  
11991     huanglatoya@example.net             CE        162723607  Inactivo  
11994         jacob96@example.com            NIT        176580896  Inactivo  

[5487 rows x 15 columns]

### 2.5 Analyze the field: 'tipo_documento' and 'numero_documento'

In [12]:
# Remove every row where 'numero_documento' is NaN or empty
df_mascotas_propietarios_cleaned = df_mascotas_propietarios_cleaned.dropna(subset=["numero_documento"])
df_mascotas_propietarios_cleaned = df_mascotas_propietarios_cleaned[df_mascotas_propietarios_cleaned["numero_documento"] != ""]
df_mascotas_propietarios_cleaned

nombre_mascota                raza   peso fecha_nacimiento    sexo  \
0        THANOS URREGO              Poodle    NaN       2022-09-01   Macho   
1       TINENK ORDONEZ             Shit-zu    NaN       2017-02-17  Hembra   
3         SUSSY PUERTA  YORK SHIRE TERRIER   3.60       2016-10-07  Hembra   
4        EEVEE ANACONA              Poodle    NaN       2017-09-03  Hembra   
5      DULCE RODRIGUEZ            PINSCHER    NaN       2020-04-01   Macho   
...                ...                 ...    ...              ...     ...   
11987          William              Poodle  11.05       2017-04-12  Hembra   
11988             Anna    Golden Retriever   5.08       2024-10-28   Macho   
11990           Andrew     German Shepherd  25.12       2016-08-23   Macho   
11991          Charles              Poodle  47.89       2017-12-10  Hembra   
11994         Jennifer    Golden Retriever  46.10       2024-07-17  Hembra   

      temperamento  color   nombre_propietario ciudad  \
0           social   None        LIZETH URREGO   Cali   
1           social   None        PAOLA ORDONEZ   Cali   
3           social   None  MARIA CAMILA PUERTA   Cali   
4           social   None       YAZMIN ANACONA   Cali   
5           social   None      JAIME RODRIGUEZ   Cali   
...            ...    ...                  ...    ...   
11987     Juguetón  Negro         Angela Moore   Cali   
11988    Tranquilo  Negro       Steven Douglas   Cali   
11990    Tranquilo   Gris         Debbie Smith   Cali   
11991       Tímido   Gris            Jose Reid   Cali   
11994     Agresivo  Negro        Steven Keller   Cali   

                                               direccion    telefono  \
0      CRA100 #28-68 VALLE DEL LILI  MADEIRO TORRE 1 ...  3207201137   
1                              cra 85 c # 33- 40 casa 54  3058147632   
3                   Cra. 98b #34-53 GUADALQUIVIR CASA 64  3134024437   
4                      Calle 45#98B-65 Apto 403, torre 8  3216066041   
5                          CRA98B#48-127  VALLE DEL LILI  3175175266   
...                                                  ...         ...   
11987                                  0106 Dudley Route  3143111203   
11988                       6458 Rogers Courts Suite 080  3191768598   
11990                            80840 Sellers Junctions  3118659393   
11991                              7585 Rodriguez Bypass  3024903661   
11994                                   2256 Daniel Camp  3139761188   

                            email tipo_documento numero_documento    estado  
0      lizethurrego1990@gmail.com             CC       1036637677    Activo  
1          pa_ordonez@hotmail.com             CC         66987417    Activo  
3                            None             CC       1006107262    Activo  
4            jazmin.dag@gmail.com             CC       1061750508    Activo  
5           leidyjoroma@gmail.com             CC          2569816    Activo  
...                           ...            ...              ...       ...  
11987  hardingkenneth@example.com            NIT        125606839  Inactivo  
11988     jeremyhuber@example.net             CE        176912006    Activo  
11990       brandon31@example.net             CC        195453083  Inactivo  
11991     huanglatoya@example.net             CE        162723607  Inactivo  
11994         jacob96@example.com            NIT        176580896  Inactivo  

[4726 rows x 15 columns]

In [13]:
# Show all the different values in 'tipo_documento'
df_mascotas_propietarios_cleaned["tipo_documento"].unique()


array(['CC', 'DII', 'Pasaporte', 'NIT', 'CE'], dtype=object)

For this analysis, we are only interested in the overage people with 'tipo documento' equals to 'CC'.

In [14]:
# Remove every row where 'tipo_documento' is not 'CC' and overage people so, fecha_nacimiento is less than 18 years
df_mascotas_propietarios_cleaned = df_mascotas_propietarios_cleaned[df_mascotas_propietarios_cleaned["tipo_documento"] == "CC"]
df_mascotas_propietarios_cleaned

nombre_mascota                raza   peso fecha_nacimiento    sexo  \
0        THANOS URREGO              Poodle    NaN       2022-09-01   Macho   
1       TINENK ORDONEZ             Shit-zu    NaN       2017-02-17  Hembra   
3         SUSSY PUERTA  YORK SHIRE TERRIER   3.60       2016-10-07  Hembra   
4        EEVEE ANACONA              Poodle    NaN       2017-09-03  Hembra   
5      DULCE RODRIGUEZ            PINSCHER    NaN       2020-04-01   Macho   
...                ...                 ...    ...              ...     ...   
11960          Jeffrey    Golden Retriever   3.81       2016-06-29   Macho   
11961           Steven             Bulldog  20.44       2018-11-02   Macho   
11966              Amy     German Shepherd  40.07       2018-12-26  Hembra   
11974           Ashlee    Golden Retriever   9.78       2016-01-01  Hembra   
11990           Andrew     German Shepherd  25.12       2016-08-23   Macho   

      temperamento   color   nombre_propietario ciudad  \
0           social    None        LIZETH URREGO   Cali   
1           social    None        PAOLA ORDONEZ   Cali   
3           social    None  MARIA CAMILA PUERTA   Cali   
4           social    None       YAZMIN ANACONA   Cali   
5           social    None      JAIME RODRIGUEZ   Cali   
...            ...     ...                  ...    ...   
11960     Agresivo  Blanco         Shannon Bean   Cali   
11961    Tranquilo   Beige   Stephanie Morrison   Cali   
11966    Tranquilo   Negro    Mr. Timothy Ayala   Cali   
11974     Agresivo   Negro        Glenda Castro   Cali   
11990    Tranquilo    Gris         Debbie Smith   Cali   

                                               direccion    telefono  \
0      CRA100 #28-68 VALLE DEL LILI  MADEIRO TORRE 1 ...  3207201137   
1                              cra 85 c # 33- 40 casa 54  3058147632   
3                   Cra. 98b #34-53 GUADALQUIVIR CASA 64  3134024437   
4                      Calle 45#98B-65 Apto 403, torre 8  3216066041   
5                          CRA98B#48-127  VALLE DEL LILI  3175175266   
...                                                  ...         ...   
11960                        509 Fox Throughway Apt. 035  3185732421   
11961                        48357 Neal Hollow Suite 405  3072605930   
11966                              18058 Richmond Fields  3156669785   
11974                                  41155 Curry Alley  3146524137   
11990                            80840 Sellers Junctions  3118659393   

                            email tipo_documento numero_documento    estado  
0      lizethurrego1990@gmail.com             CC       1036637677    Activo  
1          pa_ordonez@hotmail.com             CC         66987417    Activo  
3                            None             CC       1006107262    Activo  
4            jazmin.dag@gmail.com             CC       1061750508    Activo  
5           leidyjoroma@gmail.com             CC          2569816    Activo  
...                           ...            ...              ...       ...  
11960         pcampos@example.com             CC        183253281    Activo  
11961       zgonzalez@example.com             CC        161320236  Inactivo  
11966      hooperomar@example.net             CC        124462449    Activo  
11974       matthew86@example.org             CC        142758190  Inactivo  
11990       brandon31@example.net             CC        195453083  Inactivo  

[1749 rows x 15 columns]

#### 3. Create a merge between the new df_mascotas_propietarios_cleaned and the table: propietarios_transacciones_staging

In [15]:
# Create a merge between the new df_mascotas_propietarios_cleaned and the table: propietarios_transacciones_staging
df_propietarios_transacciones_staging = pd.read_sql('SELECT * FROM propietarios_transacciones_staging;', db_staging_engine)
df_propietarios_transacciones_staging

nombre_propietario tipo_documento numero_documento    nombre_mascota  \
0       NICOLAS LOPEZ CANO             CC       1116733904  TOMAS LOPEZ CANO   
1       NICOLAS LOPEZ CANO             CC       1116733904  TOMAS LOPEZ CANO   
2       NICOLAS LOPEZ CANO             CC       1116733904  TOMAS LOPEZ CANO   
3       NICOLAS LOPEZ CANO             CC       1116733904  TOMAS LOPEZ CANO   
4       NICOLAS LOPEZ CANO             CC       1116733904  TOMAS LOPEZ CANO   
...                    ...            ...              ...               ...   
182975       Deborah Lopez             CC        146255830           Timothy   
182976       Deborah Lopez             CC        146255830           Timothy   
182977       Deborah Lopez             CC        146255830           Timothy   
182978       Deborah Lopez             CC        146255830           Timothy   
182979       Deborah Lopez             CC        146255830           Timothy   

           servico_prestado  valor_servicio fecha_servicio  
0                Vacunación       1406023.0     2017-11-22  
1                     Otros        125698.0     2019-10-06  
2                Peluquería       1815514.0     2019-05-18  
3                     Otros       1592900.0     2023-03-20  
4                Vacunación        147163.0     2022-10-08  
...                     ...             ...            ...  
182975                 Baño        876549.0     2023-04-23  
182976      Venta Alimentos       1991466.0     2024-06-08  
182977           Peluquería       1983390.0     2023-09-11  
182978  Cita Veterinario(a)        530959.0     2024-03-10  
182979                 Baño       1712207.0     2022-12-17  

[182980 rows x 7 columns]

In [18]:
# Merge the DataFrames
df_merged = pd.merge(
    df_mascotas_propietarios_cleaned,
    df_propietarios_transacciones_staging,
    how='inner',
    left_on='numero_documento',
    right_on='numero_documento'
)
df_merged

nombre_mascota_x             raza   peso fecha_nacimiento    sexo  \
0        THANOS URREGO           Poodle    NaN       2022-09-01   Macho   
1       TINENK ORDONEZ          Shit-zu    NaN       2017-02-17  Hembra   
2       TINENK ORDONEZ          Shit-zu    NaN       2017-02-17  Hembra   
3       TINENK ORDONEZ          Shit-zu    NaN       2017-02-17  Hembra   
4       TINENK ORDONEZ          Shit-zu    NaN       2017-02-17  Hembra   
...                ...              ...    ...              ...     ...   
33397           Andrew  German Shepherd  25.12       2016-08-23   Macho   
33398           Andrew  German Shepherd  25.12       2016-08-23   Macho   
33399           Andrew  German Shepherd  25.12       2016-08-23   Macho   
33400           Andrew  German Shepherd  25.12       2016-08-23   Macho   
33401           Andrew  German Shepherd  25.12       2016-08-23   Macho   

      temperamento color nombre_propietario_x ciudad  \
0           social  None        LIZETH URREGO   Cali   
1           social  None        PAOLA ORDONEZ   Cali   
2           social  None        PAOLA ORDONEZ   Cali   
3           social  None        PAOLA ORDONEZ   Cali   
4           social  None        PAOLA ORDONEZ   Cali   
...            ...   ...                  ...    ...   
33397    Tranquilo  Gris         Debbie Smith   Cali   
33398    Tranquilo  Gris         Debbie Smith   Cali   
33399    Tranquilo  Gris         Debbie Smith   Cali   
33400    Tranquilo  Gris         Debbie Smith   Cali   
33401    Tranquilo  Gris         Debbie Smith   Cali   

                                               direccion  ...  \
0      CRA100 #28-68 VALLE DEL LILI  MADEIRO TORRE 1 ...  ...   
1                              cra 85 c # 33- 40 casa 54  ...   
2                              cra 85 c # 33- 40 casa 54  ...   
3                              cra 85 c # 33- 40 casa 54  ...   
4                              cra 85 c # 33- 40 casa 54  ...   
...                                                  ...  ...   
33397                            80840 Sellers Junctions  ...   
33398                            80840 Sellers Junctions  ...   
33399                            80840 Sellers Junctions  ...   
33400                            80840 Sellers Junctions  ...   
33401                            80840 Sellers Junctions  ...   

                            email tipo_documento_x numero_documento    estado  \
0      lizethurrego1990@gmail.com               CC       1036637677    Activo   
1          pa_ordonez@hotmail.com               CC         66987417    Activo   
2          pa_ordonez@hotmail.com               CC         66987417    Activo   
3          pa_ordonez@hotmail.com               CC         66987417    Activo   
4          pa_ordonez@hotmail.com               CC         66987417    Activo   
...                           ...              ...              ...       ...   
33397       brandon31@example.net               CC        195453083  Inactivo   
33398       brandon31@example.net               CC        195453083  Inactivo   
33399       brandon31@example.net               CC        195453083  Inactivo   
33400       brandon31@example.net               CC        195453083  Inactivo   
33401       brandon31@example.net               CC        195453083  Inactivo   

      nombre_propietario_y tipo_documento_y nombre_mascota_y servico_prestado  \
0            LIZETH URREGO               CC    THANOS URREGO        Eutanasia   
1            PAOLA ORDONEZ               CC   TINENK ORDONEZ    Desparacitada   
2            PAOLA ORDONEZ               CC   TINENK ORDONEZ    Desparacitada   
3            PAOLA ORDONEZ               CC   TINENK ORDONEZ  Venta Alimentos   
4            PAOLA ORDONEZ               CC   TINENK ORDONEZ       Vacunación   
...                    ...              ...              ...              ...   
33397         Debbie Smith               CC           Andrew  Venta Alimentos   
33398         Debbie Smith    

## Note: Now, we are going to work with a new data frame with 33402 rows with better data, instead that with 182980 rows.